tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java

```diff
@@ -0,0 +1,520 @@
+			XHTMLContentHandler xhtml) throws IOException, SAXException {
+		// BodyText 읽기
+		Entry bodyText = root.getEntry("BodyText");
+		if (bodyText == null || !bodyText.isDirectoryEntry())
+			throw new IOException("Invalid BodyText");
+
+		Iterator<Entry> iterator = ((DirectoryEntry) bodyText).getEntries();
+		while (iterator.hasNext()) {
+			Entry entry = iterator.next();
+			if (entry.getName().startsWith("Section")
+					&& entry instanceof DocumentEntry) {
+				log.debug("extract {}", entry.getName());
+
+				InputStream input = new DocumentInputStream(
+						(DocumentEntry) entry);
+				if (header.compressed)
+					input = new InflaterInputStream(input, new Inflater(true));
+
+				HwpStreamReader reader = new HwpStreamReader(input);
+
+				parse(reader, xhtml);
+				
+			} else {
+				log.warn("Unknown Entry '{}'({})", entry.getName(), entry);
+			}
+		}
+	}
+
+	/**
+	 * 텍스트 추출
+	 * 
+	 * @param writer
+	 * @param source
+	 * 
+	 * @return
+	 * @throws IOException
+	 */
+	private void parseViewText(FileHeader header, DirectoryNode root,
+			XHTMLContentHandler xhtml) throws IOException {
+		// read BodyText
+		Entry bodyText = root.getEntry("ViewText");
+		if (bodyText == null || !bodyText.isDirectoryEntry())
+			throw new IOException("Invalid ViewText");
+
+		Iterator<Entry> iterator = ((DirectoryEntry) bodyText).getEntries();
+		while (iterator.hasNext()) {
+			Entry entry = iterator.next();
+			if (entry.getName().startsWith("Section")
+					&& entry instanceof DocumentEntry) {
+				log.debug("extract {}", entry.getName());
+
+				InputStream input = new DocumentInputStream(
+						(DocumentEntry) entry);
+	
+				Key key = readKey(input);
+				try {
+					input = createDecryptStream(input, key);
+					if (header.compressed)
+						input = new InflaterInputStream(input, new Inflater(
+								true));
+
+					HwpStreamReader sectionStream = new HwpStreamReader(input);
+					parse(sectionStream, xhtml);
+				} catch (InvalidKeyException e) {
+					throw new IOException(e);
+				} catch (NoSuchAlgorithmException e) {
+					throw new IOException(e);
+				} catch (NoSuchPaddingException e) {
+					throw new IOException(e);
+				} catch (SAXException e) {
+					throw new IOException(e);
+				} finally {
+					try {
+						input.close();
+					} catch (IOException e) {
+						log.error("never happen?", e);
+					}
+				}
+			} else {
+				log.warn("unknown Entry '{}'({})", entry.getName(), entry);
+			}
+		}
+	}
+
+	private Key readKey(InputStream input) throws IOException {
+		byte[] data = new byte[260];
+
+		input.read(data, 0, 4); // TAG,
+
+		input.read(data, 0, 256);
+
+		SRand srand = new SRand(LittleEndian.getInt(data));
+		byte xor = 0;
+		for (int i = 0, n = 0; i < 256; i++, n--) {
+			if (n == 0) {
+				xor = (byte) (srand.rand() & 0xFF);
+				n = (int) ((srand.rand() & 0xF) + 1);
+			}
+			if (i >= 4) {
+				data[i] = (byte) ((data[i]) ^ (xor));
+			}
+		}
+
+		int offset = 4 + (data[0] & 0xF);

```

Comment:
add sanity/range check to offset

Reply:
I think sanity/range check to offset is not necessary.  data[] has the fixed length of 260 and offset has the range of 4 ~ 19

Reply:
doh...sorry....thank you!